### ISSUE TITLE
The SDL3 Documentation Megathread

### ISSUE THREAD

**Issue by icculus** [libsdl-org/SDL#9440]
Ok, so one thing we'd like to do as SDL3 is getting closer is improve documentation.

Documentation is a never-ending job: it will never be complete and it can always improve, so we should probably decide on things we definitely want to do _as specific tasks we can finish_. Maybe we choose new tasks later, but let's have something we can list in a milestone.

This thread is open to conversation! I'll edit this text with a checklist when we make some decisions.

**Comment by icculus** [comment]
Some things I'm planning to do in the near-term:

- Make the wiki bridge support `#defines`, enums and structs, instead of just functions. That's going to go a long way, as far as formal documentation is concerned.
- Make a YouTube video where I upgrade [DirkSimple](https://github.com/icculus/DirkSimple) from SDL2 to SDL3.
- Go through all the SDL3 wiki pages and clean out obsolete stuff: move everything to Markdown, delete pages for APIs that were removed/renamed, SDL2-specific pages, the old migration guide. SDL2's subdir I'm probably going to leave alone, but it could probably stand some weeding too.
- We have to decide if we want to write a less-formal migration guide. Right now I'm on the fence about it, to be honest.

**Comment by slouken** [comment]
Can you add documentation for hints as well?

**Comment by icculus** [comment]
> Can you add wiki bridge support for hints as well?

I think they'll probably come with the `#define` support, but if not, yes, we can add that.

**Comment by icculus** [comment]
> Make a YouTube video where I upgrade [DirkSimple](https://github.com/icculus/DirkSimple) from SDL2 to SDL3.

Something like ioquake3 might be cool too, while I'm thinking about it.

**Comment by slouken** [comment]
Yeah, something people can actually build and run without ripping a laserdisk might be nice. :)

**Comment by Svintaj** [comment]
I would like to see a tutorial on a **'recommended cross-platform workflow'**. 

That show how to build a basic SDL3 example to Linux, Windows, macOS, Android and iOS. 
So a new SDL3 user can have a smooth working cross-platform workflow from start. 

I think there is more people like me that know some C++ but not much about cross-platform compiling. I guess you, the SDL3 developers are kind of experts on **cross-platform workflows** and could give us some good advice, for a smooth path forward. 👋

**Comment by icculus** [comment]
> I would like to see a tutorial on a 'recommended cross-platform workflow'.

Strictly speaking, this is what https://github.com/libsdl-org/SDL_helloworld is for, but it could use some more love.

**Comment by Svintaj** [comment]
>it could use some more love

Ok, yes please, for me it is still like a riddle to decipher. 🤔
I would like more explanations on **how** and **why**. Preferably in a video tutorial! 🌞

I got a bunch of basic questions like: What am I supposed to do with this Hello World repository, to build main.c for different platforms than the one I’m currently on? What is ‘CI’ or ‘GitHub Actions’ and are they essential for cross-platform compiling? What shall I do with the .yml file? How should I modify the CMakeLists.txt file. Etc.

In the tutorial, pretend to explain the process for someone that only have learned plain C or C++ through a book. -> What more do someone need to learn in that case, to be able to cross-compile? 

Since it’s a Hello World example for a Cross-Platform lib, I think it’s worth explaining it from the ground up. 🤓👍

**Comment by icculus** [comment]
Okay, I'll see if I can make a quick video for that!

**Comment by icculus** [comment]
> I'm still tweaking the code, but now this is also causing a big task of cleaning up stuff that was already written for the wiki for some of these things. I'm merging documentation where it's reasonable, throwing a lot of it out as obsolete or just bad, and converting from mediawiki to markdown format.

The first wave of this is done now. I'm going to push this in the morning, so I'm alert in case it blows up the wiki server or something.

**Comment by madebr** [comment]
Since SDL3, CMake installs man pages.
These man pages are generated using the perl script.
Because the wikiheaders project is not available, the man pages don't contain eventual examples present in the wiki.
Perhaps the bridge should also transfer example code to the SDL project?

**Comment by icculus** [comment]
This is intentional since it clutters up the headers, and honestly most example code you'd write for a single function is very close to useless, imho.

**Comment by madebr** [comment]
That's a valid point, and I agree.
I was thinking about storing it somewhere in structured (testable) form

**Comment by icculus** [comment]
Generic typedefs (if they have a Doxygen-style comment right above them, mind you) are bridged to the wiki now. A lot of these are useful in the name of completeness, but it's _super_ useful for things like this:

https://wiki.libsdl.org/SDL3/SDL_AudioStreamCallback

**Comment by madebr** [comment]
Can your documentation bot also synchronize `wikiheaders.pl` across the satellite libraries?
e.g. [`build-scripts/wikiheaders.pl`](https://github.com/libsdl-org/SDL_image/blob/main/build-scripts/wikiheaders.pl) for SDL_image

**Comment by icculus** [comment]
Probably. I'll take a look.

**Comment by Sackzement** [comment]
Just wanted yo throw this in here.
This is my reference for SDL.
Maybe it can give you some ideas.
[SDL3.2.0 Reference.ods](https://github.com/libsdl-org/SDL/files/14967272/SDL3.2.0.Reference.ods)

**Comment by icculus** [comment]
> Can your documentation bot also synchronize wikiheaders.pl across the satellite libraries?

Against my better judgement, I took a run at SDL2's docs so it can use the latest wikiheaders.pl too. This was tedious, but resulted in cleaning out a _ton_ of garbage.

More or less, I did a much faster version of the cleanups I already did in SDL3 (hours, not days). I've likely made some small mistakes, but hopefully this improves things from what was there.

But if anyone sees something obviously broken in the SDL2 headers or wiki, please either tell me or just push a fix.

**Comment by Sackzement** [comment]
An entry in the Wiki about which C standard is used by SDL could be useful.
Something like:
> ### C Standard:
> Building a project with SDL: C89 and newer
> Building the SDL library itself: C99 and newer

And maybe a note about which C99 features are used to build SDL itself:
> ### C99 features used to build the SDL library itself:
> // comments
> variable declarations in for loop initializers
> va_copy function from <stdarg.h> (src/SDL_log.c)
> inline function specifier (in dbus-1.0 headers)

**Comment by icculus** [comment]
I added a little code to wikiheaders to tell me what we _aren't_ covering right now. The output's not elegant, but it gets the job done.

I expected this to be a lot of things that the script doesn't document, but it turns out, we can turn almost any piece of the SDL headers into a wiki page at this point, which is pretty wild.

So this ended up being a useful way to see what things are raw text that either have broken comments or lack documentation.

For those inclined to sift through it, I put a trimmed-down version in a gist:

https://gist.github.com/icculus/565ba73ad134ede12905b26792963693

Some of this can be ignored (`\file` comment blocks, include-once macro stuff, other preprocessor salsa), but it's pretty clearly shows what needs fixing up in the headers still. Not a lot, but there are a few things.

But the tl;dr on this: the wiki/header documentation work is basically almost done at this point. We can always improve existing documentation, but there is very little that isn't documented _at all_ now!

**Comment by madebr** [comment]
This is bike shedding, but I see this warning message when generaring the SDL man pages:
```
  Use of uninitialized value $brief in string at /var/folders/3m/p59k4qdj0f17st0gn2cmj3640000gn/T/setupsdl/66c57facf1111a4ec08d3d1abdf3c87f3062ffbea8d9eb4c9c412e9b5e7f59b7/source/build-scripts/wikiheaders.pl line 1902.
  Use of uninitialized value $brief in string at /var/folders/3m/p59k4qdj0f17st0gn2cmj3640000gn/T/setupsdl/66c57facf1111a4ec08d3d1abdf3c87f3062ffbea8d9eb4c9c412e9b5e7f59b7/source/build-scripts/wikiheaders.pl line 1902.
```
[ci link](https://github.com/madebr/SDL_image/actions/runs/8917448500/job/24490526549#step:5:960)

**Comment by icculus** [comment]
Manually cleaned out that gist so it's only things that are at least worth considering (and maybe ignoring), or need simple fixes. The most obvious things that could do better in wikiheaders itself are documentation for properties (which I'm inclined to group with some other page, as we currently do for `#define`s right below a typedef), and the `\file` comment blocks, which honestly should bridge to various Category pages on the wiki.

I'll look at the perl warning in the morning, it looks like it's only macOS's perl doing it, but it's probably a legit bug in my code.

Also: this came up today: we shouldn't generate the manpages by default. Most people don't need them, and they add time and requirements to the build process. Plus, they require a case-sensitive filesystem, since at a minimum we have unrelated `SDL_Log` and `SDL_log` functions.

Let's make it easy for the things that need them (package maintainers, the true Unix devotees, and GitHub Actions) to turn them on with a CMake variable, but let's default building manpages to off.

**Comment by madebr** [comment]
Makes sense. I'll do that.

**Comment by icculus** [comment]
Properties now go to the wiki: https://wiki.libsdl.org/SDL3/SDL_CreateWindowWithProperties

**Comment by slouken** [comment]
> Properties now go to the wiki: https://wiki.libsdl.org/SDL3/SDL_CreateWindowWithProperties

I suggest not including all the `#define` lines. Applications should be using the names of the properties, not the raw strings. Also maybe we don't need to create individual files for each property, unless that's useful for search or something?

**Comment by icculus** [comment]
> I suggest not including all the #define lines.

I was torn about that, but I am absolutely certain all these scripting language bindings that are banging on our door all the time will appreciate knowing the actual string.

(and we can't _change_ the strings later anyhow, since it would break ABI, so if someone ignores the macro and uses the string...)

> unless that's useful for search or something?

They will come up in a search on the wiki itself (which is useful if the function docs don't directly mention it), but also I think it's nice for people to be able to say "what the heck is this thing?" and either punch it into the address bar on wiki.libsdl.org, or find that page on a Google search, without us having to write repetitive documentation for them.

**Comment by icculus** [comment]
Updated the gist; the remaining documentation gaps are:

- function pointer typedefs that are split over multiple lines are currently ignored and shouldn't be.
- the main callbacks are documented but the SDLMAIN_DECLSPEC symbol makes them get ignored by the script.
- SDL_stdinc.h, which is _daunting_ if we want to get into it, which maybe we don't right now.
- The SDL_THREAD_ANNOTATION_ATTRIBUTE stuff, which I think we'll leave undocumented.
- SDL_PixelFormatEnum and SDL_Colorspace, which are gigantic but don't need a lot of documentation.
- Bridging the `\file` stuff to the Category pages.
- Things like SDL_NORETURN and SDL_FORCE_INLINE in SDL_begin_code. Less important, not sure we intend apps to be using these directly anyhow.
- SDL_intrin.h probably doesn't need all the `SDL_SSE_INTRINSICS` macros documented, same reason.
- SDL_platform_defines.h, which is probably not worth documenting platform-by-platform.
- SDL_pen.h needs several adjustments from `/**< one liners */` to documentation above the thing.
- A couple other legitimate things here and there, like `SDL_IOStatus`, etc.

I can feel myself petering out on this work, so I'm going to try to wrap this up tomorrow and get back to other tasks. After that, we should leave this issue open for later, but I think we're at the point where we have what amounts to a pretty thorough reference manual of the entire library.

We should have wikiheaders generate LaTeX and publish it, lol.

**Comment by icculus** [comment]
> Sure, but they can look in the header, this shouldn't be on the wiki. It's verbose and redundant with the documentation below it.

Done in 77f4ffcc79fa35bb76de745f8e994ad5797cf978.

This won't show them in https://wiki.libsdl.org/SDL3/SDL_CreateWindowWithProperties, but it still generates the "go look at this function for details" pages, like this one: https://wiki.libsdl.org/SDL3/SDL_PROP_WINDOW_CREATE_FOCUSABLE_BOOLEAN

**Comment by slouken** [comment]
Perfect, thanks!

**Comment by icculus** [comment]
> We should have wikiheaders generate LaTeX and publish it, lol.

I was joking when I said this, but I have to say...I don't hate it.

[test.pdf](https://github.com/libsdl-org/SDL/files/15204148/test.pdf)

**Comment by icculus** [comment]
The test.pdf was just one of the wiki pages pushed through pandoc, but I spent a little time to make wikiheaders actually spit out LaTeX directly, and it was informative:

[SDL-3.1.2-reference-manual.pdf](https://github.com/libsdl-org/SDL/files/15211989/SDL-3.1.2-reference-manual.pdf)

First, this could never go forward because it would need an enormous amount of manual labor to make it friendly to a book's dimensions. A lot of it works, but there are too many places where preformatted text and code blocks blow up and run off the end of the paper. I suspect fixing this isn't going to be a script thing, it's going to be hand-editing documentation to be less nice for the places we actually care about it: the wiki, the headers.

Second, we could probably format this more tightly with some different parameters, but holy moly, it's a _SIXTEEN HUNDRED PAGE BOOK_. No one is ever going to print this. Or lift it. With just the functions and no macros or datatypes, it's still like 1200 pages as-is.

(The PDF is still pretty cool, though.)

**Comment by icculus** [comment]
Now-bogus pages for removed/renamed APIs have been removed from the wiki (94 of them, total). All the old MoinMoin syntax is gone now, too, I think.

SDL2's wiki is still a disaster. Just so much bitrot to clean out of there, but I don't know if anyone will ever do it.

**Comment by slouken** [comment]
> SDL2's wiki is still a disaster. Just so much bitrot to clean out of there, but I don't know if anyone will ever do it.

Nope, let's focus on SDL3. :)

**Comment by icculus** [comment]
> Nope, let's focus on SDL3. :)

Agreed. I cleaned out some obvious messes real quickly, but there is a _ton_ of documentation we wrote for SDL3 that would be an signficant effort to reproduce for SDL2.

That being said: I'm sorting through all the open issues on sdlwiki, and almost _all_ of them so far were already fixed by wikiheaders, since it cleaned out a bunch of weird things and made everything look uniform. Most of the rest of the issues _will be_ fixed when I made it automatically categorize pages by subsystem.

So I think SDL2's docs will have gaps (because no one wrote some pieces at all), but the wiki won't be unusable.

**Comment by icculus** [comment]
Also discovering from the unmerged sdlwiki pull requests how much duplicated labor we're avoiding with wikiheaders going forward. This was definitely a win.

**Comment by icculus** [comment]
Okay, cleaned out most of the sdlwiki issues and pull requests. There's a few still open I don't know how to respond to yet, but that's like a 2-year backlog cleared out.

It showed that category management is clearly the biggest pain point, both for editors and readers, and also the one most obviously fixable by adding it to wikiheaders, so that's my next task.

**Comment by icculus** [comment]
Wrote up some documentation on how to write documentation. It just sort of explains the quirks wikiheaders uses to make things work, so when things break I can point to a section to explain why.

https://wiki.libsdl.org/SDL3/README/documentation-rules

(or `docs/README-documentation-rules.md`)

**Comment by slouken** [comment]
> Wrote up some documentation on how to write documentation. It just sort of explains the quirks wikiheaders uses to make things work, so when things break I can point to a section to explain why.
> 
> https://wiki.libsdl.org/SDL3/README/documentation-rules
> 
> (or `docs/README-documentation-rules.md`)

Wow, this is really good documentation documentation. Thanks!

**Comment by icculus** [comment]
Ran across this, worth considering:

https://blog.theprogrammingjunkie.com/post/sdl2-cheatsheet/

**Comment by icculus** [comment]
Okay, the category documentation work is done!

Now people might start at...

https://wiki.libsdl.org/SDL3/APIByCategory

...and find the audio section, clicking through to...

https://wiki.libsdl.org/SDL3/CategoryAudio

Where they'll get a quick overview, and then a list of everything in the category. This should make it easier to navigate through to what they need.

And of course, it's bridged to the headers.  :)  For the most part, one header is one category on the wiki, so the mental grouping is the same.

A _lot_ of these categories need documentation written, as most have one-sentence placeholders. But all the code and cleanup is in place now.

This was the last big pending task on wikiheaders, fwiw.   :)

**Comment by madebr** [comment]
Nice job!

Is the 2px padding required on `code` items? That makes the first line to be indendented a bit.
For example, see the code block at https://wiki.libsdl.org/SDL3/CategoryAudio

(My browser is firefox)

**Comment by icculus** [comment]
That's probably a CSS bug. We can fix that, I assume.

**Comment by icculus** [comment]
Pushed a fix for the CSS. Note that firefox caches the CSS for the wiki _really_ aggressively, and I'm not sure why. I ended up hitting F12, finding the "network" tab and clicking "disable cache" before reloading the page to get it to catch up...but with the right CSS, it's definitely fixed here.

**Comment by madebr** [comment]
I think caching can be sidestepped by adding (dummy) query strings to the url.
e.g. [https://wiki.libsdl.org/static_files/ghwikipp.css?h=026e7d33](https://wiki.libsdl.org/static_files/ghwikipp.css?h=026e7d33)
And calculate the query as the md5 hash of  the file contents.

**Comment by icculus** [comment]
Example code is being discussed over in #10350.

**Comment by icculus** [comment]
Wikiheaders should probably make pages for each item in an enum, so when someone says "hey, what does SDL_MESSAGEBOX_COLOR_BUTTON_SELECTED mean?" they'll get a page that says "please refer to SDL_MessageBoxColorType for details" with a clickable link.

We already do this for defines that follow a typedef, to catch bitflags.

**Comment by ritalat** [comment]
I feel like the _building an app_ section of the Android readme could use some work. I was trying to port a toy project to Android out of curiosity but found the instructions messy and invasive to an existing project with all the symlinking/copying sources and assets around. It makes it seem like you need to restructure your project just to saitsfy Android when it really isn't necessary.

Now admittedly I've never touched Android development before so I'm probably unaware of a lot of basics but I did end up taking notes of the steps that worked for me:

```
ANDROID PORTING
===============
let's say our existing example project looks something like this:
MyProject
  -assets
  -SDL (git submodule etc.)
  -shaders
  -src

1. copy the contents of 'android-project' into your project directory
2. edit app/build.gradle to include your assets by adding assets.srcDirs in sourceSets.main
  assets.srcDirs = ['../assets', '../shaders']

IF USING CMAKE
3c. set buildWithCMake to true and set externalNativeBuild cmake path to your own top level CMakeLists.txt
4c. change your main executable to be built as a library instead when targeting android

IF USING ANDROID MAKEFILES
3m. copy app/jni/src/Android.mk to your source directory and add your sources to it
4m. replace the contents of app/jni/Android.mk with includes to SDL's Android.mk and your source Android.mk you just copied and edited
  include ../SDL/Android.mk
  include ../src/Android.mk

FINALLY
5. run gradlew or open the project in android studio!
```

This way I didn't have to restructure the existing project at all. You just end up with a new 'app' subdirectory and a bunch of gradle files.

**Comment by icculus** [comment]
Put together a video on moving ioquake3 from SDL2 to SDL3 last night.

https://youtu.be/i3yVqWYFbCE

It's 90 minutes long, so I don't expect anyone to plow through this in one sitting, but it actually worked really well in terms of having good examples for a lot of the important points.

It's unlisted right now (outside of this github comment), but I'm going to publish it for real shortly.

**Comment by madebr** [comment]
Should we put the coccinelle script in the development packages?

**Comment by Sackzement** [comment]
> Put together a video on moving ioquake3 from SDL2 to SDL3 last night.
> 
> https://youtu.be/i3yVqWYFbCE
> 
> It's 90 minutes long, so I don't expect anyone to plow through this in one sitting, but it actually worked really well in terms of having good examples for a lot of the important points.
> 
> It's unlisted right now (outside of this github comment), but I'm going to publish it for real shortly.

Really good video.
I like, that you explained in more detail why changes were made and the design around it, alongside what changed.

You just should have made a single kill at the end :)

**Comment by icculus** [comment]
> Are the pages like the mentioned [keyboard practices page](https://wiki.libsdl.org/SDL3/BestKeyboardPractices) hyperlinked from somewhere?

Honestly, I don't think it is at the moment. When Sam and I were iterating on how to handle keyboard input, we kept coming back to "these are the four ways an app might want to get keyboard input," so I wrote it out as we talked through the API decisions...but I don't know where we should put it to make it findable.

Definitely in [CategoryKeyboard](https://wiki.libsdl.org/SDL3/CategoryKeyboard), in the not-yet-written intro section. But other suggestions are welcome.

**Comment by icculus** [comment]
That's a good idea!

**Comment by icculus** [comment]
Here's the DirkSimple SDL2->SDL3 video I threatened to make:

https://youtu.be/V_MWXDr3Nu4

It is _also_ 90 minutes, 35 of them the basic port, and the rest is "let's replace some existing annoyances with new SDL3 features," which I think makes the video a nice value-add over the ioquake3 one.

**Comment by icculus** [comment]
I took a run at having wikiheaders generate a "quick reference" doc, in the style of [this SDL2 one I mentioned](https://blog.theprogrammingjunkie.com/post/sdl2-cheatsheet/).

If we do this, it'll just be a page on the wiki that updates with new commits, so we don't have to explicitly maintain it.

https://gist.github.com/icculus/589d9add85f5a3ec772530dd58609c75

Feedback welcome.

**Comment by madebr** [comment]
That's a nice overview!
- Should the functions be ordered? Alphabetically is easy, by priority is harder (can we give functions weights?).
- `SDL_Swap16` is present multiple times
- Should every category also document the function typedefs?
 e.g. `SDL_ShowOpenFileDialog` has `SDL_DialogFileCallback` which is a function typedef

**Comment by icculus** [comment]
So for ordering, it uses the same order they exist in the headers, since I suspect that flows most logically.

SDL_Swap is probably a bug, I'll check that.

It's only doing functions, no macros. I'm worried it's too big already, so I limited it. This is also why it doesn't have typedefs; it's a quick reference to get started or refresh one's memory; they can get more detailed docs from there.

At least, I think so, but I can be persuaded otherwise.

**Comment by icculus** [comment]
Then again, the assert section is kinda worthless without SDL_assert.  :/

**Comment by icculus** [comment]
Kind of wondering if we should move wikiheaders to its own repo, and have GitHub Actions just pull the latest when building. This would solve the question of how to sync it between repos.

**Comment by slouken** [comment]
> So for ordering, it uses the same order they exist in the headers, since I suspect that flows most logically.
> 
> SDL_Swap is probably a bug, I'll check that.
> 
> It's only doing functions, no macros. I'm worried it's too big already, so I limited it. This is also why it doesn't have typedefs; it's a quick reference to get started or refresh one's memory; they can get more detailed docs from there.
> 
> At least, I think so, but I can be persuaded otherwise.

Agreed, functions and function-like macros should be listed here, types are referenced by command line parameters and don't need to be otherwise documented. Is there a way to make each of those (functions and types) link to the wiki documentation if you click on them? That seems like it would make it a very useful quick-reference sheet.

**Comment by icculus** [comment]
> Is there a way to make each of those (functions and types) link to the wiki documentation if you click on them?

Not without losing syntax highlighting, unfortunately. But they can probably copy/paste it as plain text from here into their IDE to look at it with the syntax highlighting of their choice.

**Comment by icculus** [comment]
Actually, we can probably leave this more or less as-is as a downloadable C header thing, and change the wiki to just have function names and one-liner notes that they can click through, which might be way more readable on a phone.

Just a thought.

**Comment by madebr** [comment]
> Kind of wondering if we should move wikiheaders to its own repo, and have GitHub Actions just pull the latest when building. This would solve the question of how to sync it between repos.

`wikiheaders.pl` is needed by the CMake script of all satellite libraries when configuring with `-DSDL(IMAGE|MIXER|NET|TTF)_INSTALL_MAN`, so it must be part of the repo/source archive.

In addition of `wikiheaders.pl`, there are other files that can/should be synchronized: `cmake/sdlcpu.cmake`, `build-scripts/build-release.py` and `build-scripts/create-release.py`.

**Comment by icculus** [comment]
> change the wiki to just have function names and one-liner notes that they can click through, which might be way more readable on a phone.

It's way more readable on a phone, in that it doesn't make you scroll sideways because of preformatted text going off the screen, but I sort of hate it:

https://gist.github.com/icculus/93e049c858ea12f3c0d2943d1f18fc33

Even discounting that I'm spending way too long on this now, let's just leave it as a C header-like thing, with a note about using [APIByCategory](https://wiki.libsdl.org/SDL3/APIByCategory) if they want to click around the wiki, and leave it at that.

**Comment by icculus** [comment]
> so it must be part of the repo/source archive.

...but doesn't it also have to clone libsdl-org/sdlwiki to generate manpages? So why not one more repo after that?

**Comment by icculus** [comment]
Okay, here's all the macros in the headers with documentation:

https://gist.github.com/icculus/67478f94d73970eaa6e4b37e5eade0bf

Of these, in a quick-reference thing, I'd want...

- SDL_Swap*
- SDL_assert*
- SDL_arraysize
- SDL_AtomicIncRef, SDL_AtomicDecRef
- _maybe_ the SDL_AUDIO_* macros

...and basically nothing else. So maybe we add something to .wikiheaders-options to specify macros we want in there, like

```
quickrefallowedmacros = SDL_Swap16BE,SDL_SwapLE16,SDL_assert,SDL_AtomicIncRef`
```

(etc)

and call it done.

**Comment by madebr** [comment]
> > so it must be part of the repo/source archive.
> 
> ...but doesn't it also have to clone libsdl-org/sdlwiki to generate manpages? So why not one more repo after that?

The wiki is [regenerated from scratch](https://github.com/libsdl-org/SDL/blob/48b6b93fb68ea2043254b96237c3a0ae5531f6e9/cmake/sdlmanpages.cmake#L46-L58).
The first command generates the wiki and the second command generates the actual man pages.
CMake is not interested in the wiki, but it's a required intermediate step for the wikiheaders script.

To observe it in action, configure your SDL with `-DSDL_DISABLE_INSTALL_DOCS=OFF`, and run `ninja SDL3-docs`. The files are generated in the `docs` folder.

**Comment by icculus** [comment]
Okay, quick reference is in with be5fba963a1b89a6b6c10cf62b0bf809973e1061. I made some compromises on macros, and we can adjust the list as we like by editing .wikiheaders-options for a given project.

I've turned it on for SDL_mixer, image, ttf, and net, too. Not turning it on for SDL2 itself or any SDL2 branches. It can work there, but we'd have to improve the documentation in general first, and I'm not going to spend time on that.

**Comment by icculus** [comment]
What platforms have we removed for SDL3? We should have a wiki page about this.

- OS/2
- N-Gage
- Google Stadia
- Older macOS releases

Anything else?

**Comment by madebr** [comment]
UWP is gone too

**Comment by icculus** [comment]
Oh, yeah, how could I forget?!  :)

I've updated the list in my previous comment, thanks!

**Comment by icculus** [comment]
https://wiki.libsdl.org/SDL3/SupportedPlatforms

**Comment by icculus** [comment]
As of f0fad41f2cb599aab2fd458e3c0192551c591365, SDL3 has 100% documentation coverage. Every public symbol in the headers has a wiki page now!

**Comment by slouken** [comment]
> As of [f0fad41](https://github.com/libsdl-org/SDL/commit/f0fad41f2cb599aab2fd458e3c0192551c591365), SDL3 has 100% documentation coverage. Every public symbol in the headers has a wiki page now!

Woo!!

**Comment by icculus** [comment]
Opened a new bug for the one last thing I feel is crucial for 3.2.0. The rest of this can bump to a later milestone, but honestly we're in a _really_ good place with documentation for SDL3 at this point.

**Comment by icculus** [comment]
I think we're good on this bug; the last few minor items can pop up in new bugs if necessary.
